### PR TITLE
fix #296 Compression settings don't work.

### LINKED ISSUE #296: Compression settings don't work.
### Questions

Compression settings don't work.

### Version

4.3.2

### Context

```java
        JsonObject config = new JsonObject();
        config.put("connection_string", "mongodb://root:root@127.0.0.1:37171,127.0.0.1:37172/?compressors=snappy")
            .put("useObjectId", true)
            .put("compressors", "snappy")
            .put("waitQueueTimeoutMS", Duration.ofDays(1).toMillis())
            .put("maxIdleTimeMS", 200000)
            .put("db_name", "statistic")
            .put("maxPoolSize", 6)
            .put("minPoolSize", 1);

        mongoClient = MongoClient.create(vertx, config);
        return mongoClient;
```
`compressors=snappy` does't work.

### Do you have a reproducer?

A reproducer is a simple project hosted on GitHub (or another forge supporting git clone operation) that has a build file that can be executed to reproduce the issue.

Reproducers are very helpful for contributors and will likely help them fixing your bug faster.

* Link to github project/gist

### Steps to reproduce

1. ...
2. ...
3. ...
4. ...

### Extra

* Anything that can be relevant such as OS version, JVM version

### PR DESCRIPTION
**Pull request by tarvfx**
Motivation:

fix #296 Compression settings don't work.

Conformance:

Your commits should be signed and you should have signed the Eclipse Contributor Agreement as explained in https://github.com/eclipse/vert.x/blob/master/CONTRIBUTING.md
Please also make sure you adhere to the code style guidelines: https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines

### CODE REVIEW

File: src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  private final ConnectionString connectionString;
+  private final JsonObject config;
+
+  public CompressorListParser(ConnectionString connectionString, JsonObject config) {
+    this.connectionString = connectionString;
+    this.config = config;
+
+    this.compressorKeys = new ArrayList<>();
+    this.compressorKeys.add("compressors");
+    this.compressorKeys.add("zlibcompressionlevel");
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java:23]: Please change to camel case string

File: src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import com.mongodb.MongoCompressor;
+import com.mongodb.lang.Nullable;
+import io.vertx.core.json.JsonObject;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class CompressorListParser {
+
+  private final List<String> compressorKeys;
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java:13]: It's not necessary to create at runtime a list of keys for config. Instead, we can do with a couple of `if` statements in the `compressorList` method

File: src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    } else {
+      List<MongoCompressor> compressorList = buildCompressors(compressors, zlibCompressionLevel);
+      if (connStrCompressors != null) {
+        compressorList.addAll(connStrCompressors);
+      }
+      return compressorList;
+    }
+  }
+
+  private List<MongoCompressor> buildCompressors(final String compressors, @Nullable final Integer zlibCompressionLevel) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java:55]: Don't use `@Nullable` annotations in Vert.x code please

File: src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    for (final String key : compressorKeys) {
+      String value = config.getString(key);
+      if (value == null) {
+        continue;
+      }
+
+      if (key.equals("compressors")) {
+        compressors = value;
+      } else if (key.equals("zlibcompressionlevel")) {
+        zlibCompressionLevel = Integer.parseInt(value);
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java:39]: Please catch the `NumberFormatException` and rethrow as `IllegalArgumentException` for consistency with other parsers

File: src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        compressorList.addAll(connStrCompressors);
+      }
+      return compressorList;
+    }
+  }
+
+  private List<MongoCompressor> buildCompressors(final String compressors, @Nullable final Integer zlibCompressionLevel) {
+    List<MongoCompressor> compressorsList = new ArrayList<>();
+
+    for (String cur : compressors.split(",")) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/impl/config/CompressorListParser.java:58]: We could be more tolerant and trim+lowercase each compressor name provided by users

**Review by tsegismont — CHANGES_REQUESTED** [review]
Thank you @tarvfx 

Can you please add:

- a test which shows config takes precedence over values in connection string
- documentation (in particular it should be said that the user must import a specific library for Snappy or Zstandard compression)
- a test which verifies connectivity with zlib compression

**Comment by tarvfx** [comment]
Hi, @tsegismont 

Thanks for your review and many suggestions. I'll fix these issues.

Regards

**Comment by tsegismont** [comment]
Thank you!

**Comment by tarvfx** [comment]
@tsegismont I'm Sorry, this PR has to be abolished. I will submit it again.

**Comment by tarvfx** [comment]
#298

**Pull request closed without merging**